pkg/api/v1alpha1/cloudstackmachineconfig_types.go

```diff
@@ -40,6 +45,61 @@ type CloudStackMachineConfigSpec struct {
 	UserCustomDetails map[string]string `json:"userCustomDetails,omitempty"`
 }
 
+type CloudStackResourceDiskOffering struct {
+	CloudStackResourceIdentifier `json:",inline"`
+	// path the filesystem will use to mount in VM
+	MountPath string `json:"mountPath"`
+	// device name of the disk offering in VM, shows up in lsblk command
+	Device string `json:"device"`
+	// filesystem used to mkfs in disk offering partition
+	Filesystem string `json:"filesystem"`
+	// disk label used to label disk partition
+	Label string `json:"label"`
+}
+
+func (r *CloudStackResourceDiskOffering) Equal(o *CloudStackResourceDiskOffering) bool {
+	if r == o {
+		return true
+	}
+	if r == nil || o == nil {
+		return false
+	}
+	if r.Id != o.Id {
+		return false
+	}
+
+	if r.MountPath != o.MountPath || r.Filesystem != o.Filesystem || r.Label != o.Label || r.Device != o.Device {
+		return false
+	}
+	return r.Id == "" && o.Id == "" && r.Name == o.Name
+}
+
+func (r *CloudStackResourceDiskOffering) Validate() (err error, field string, value string) {
```

Comment:
This method currently is only being called in webhook (controller). We need it in the config manager / API validation as well for CLI validation. Can we add it in https://github.com/aws/eks-anywhere/blob/main/pkg/api/v1alpha1/cloudstackmachineconfig_types.go#L98
Also curious where the other context unaware cloudstack machine config API validation are?

Reply:
The return should only include `error`, you don't need to include the key/value in webhook error

Reply:
discussed with Jia offline and reached an agreement that field name and value are needed for validateUpdate in webhook call.

In validateCreate, a simple bad request can be returned, but for validteUpdate, field name has to be specified in exception.